commit 6f22a82a401d267e4bf1fcbcff8d6adb24e14d58
Author: Michael Paquier <michael@paquier.xyz>
Date:   Sun Jul 27 17:48:47 2025 +0900

    Add assertions for all the required index AM callbacks
    
    Similar checks are done for the mandatory table AM callbacks.  A portion
    of the index AM callbacks are optional and can be NULL; the rest is
    mandatory and is documented as such in the documentation and in amapi.h.
    
    These checks are useful to detect quickly if all the mandatory callbacks
    are defined when implementing a new index access method, as the
    assertions are run when loading the AM.
    
    Author: Japin Li <japinli@hotmail.com>
    Discussion: https://postgr.es/m/ME0P300MB0445795D31CEAB92C58B41FDB651A@ME0P300MB0445.AUSP300.PROD.OUTLOOK.COM

diff --git a/src/backend/access/index/amapi.c b/src/backend/access/index/amapi.c
index f0f4f974bce..60684c53422 100644
--- a/src/backend/access/index/amapi.c
+++ b/src/backend/access/index/amapi.c
@@ -40,10 +40,23 @@ GetIndexAmRoutine(Oid amhandler)
 
 	if (routine == NULL || !IsA(routine, IndexAmRoutine))
 		elog(ERROR, "index access method handler function %u did not return an IndexAmRoutine struct",
 			 amhandler);
 
+	/* Assert that all required callbacks are present. */
+	Assert(routine->ambuild != NULL);
+	Assert(routine->ambuildempty != NULL);
+	Assert(routine->aminsert != NULL);
+	Assert(routine->ambulkdelete != NULL);
+	Assert(routine->amvacuumcleanup != NULL);
+	Assert(routine->amcostestimate != NULL);
+	Assert(routine->amoptions != NULL);
+	Assert(routine->amvalidate != NULL);
+	Assert(routine->ambeginscan != NULL);
+	Assert(routine->amrescan != NULL);
+	Assert(routine->amendscan != NULL);
+
 	return routine;
 }
 
 /*
  * GetIndexAmRoutineByAmId - look up the handler of the index access method
